README.md
# Description
Account Management tool service allows you to add, delete, retrieve account data by calling the URL with correct HTTP request method.

# Requirements
- Docker

# Deployment
Docker is our main deployment method, run the following to start your service

```
docker build -t new_app_image .
docker run -d --name new_container_name -p 80:80 new_app_image
```

Now that your service is running in the background, you can access it using `http://127.0.0.1/docs`

# Example Requests

## API Resources

  - [GET /health](#api-health)
  - [GET /accounts/[id]](#api-get-account)
  - [POST /accounts/[account_id]](#api-create-account)
  - [DELETE /accounts/[account_id]](#api-delete-account)

### [GET](#api-health) [/health]

Example: http://127.0.0.1/health

Response body:

    {
        "status":true
    }

### [GET](#api-get-account) /accounts/[account_id]

Example: http://127.0.0.1/accounts/1

Response body:

    {
        "name": "joe",
        "balance": 500,
        "active": true,
        "description": null
    }

### [POST](#api-create-account) /accounts/[account_id]

Example: http://127.0.0.1/accounts/4

Request body:

    {
        "name": "new_account",
        "description": "info",
        "balance": 1200,
        "active": true
    }

### [DELETE](#api-delete-account) /accounts/[account_id]

Example: http://127.0.0.1/accounts/1

Response body:

    {
        true
    }


# Improvements
- more functions & tests such as disabling and re-enabling accounts based on the class parameter that is available, or even querying multiple accounts at once with support for filtering.
- publishing python package as artifact and reusing it inside Dockerfile.
- implementing UI for calling the APIs which makes it easier for customers to use.
- applying rate limiting on resources based on your needs.
- having dedicated environments for your APIs that you can test before releasing to prod.
- having SSL certificates enabled on your endpoint and extra security measures such as token validation through various API manager tools.
- better error descriptions for easier use of the API.
- implement monitoring and alerting
- implement logging

CHANGELOG.md
feat:
- add API tests
- add Docker deployment

docs:
- update online auto-generated documentation - `http://127.0.0.1:8000/docs`
- add README.md

fix:
- correct `delete_account` function by writing implementation
- change account creation request from `put` to `post`. `post` is used when we create a resource, whereas `put` is used to update it. [reference](https://www.rfc-editor.org/rfc/rfc7644#section-3.5.1)

app/test.py
import main
from fastapi.testclient import TestClient

client = TestClient(main.app)

# create temporary account dictionary
main.accounts = {
    1: {"name": "joe", "balance": 500, "active": True, "description": None},
    2: {"name": "smith", "balance": 400, "active": True, "description": None},
    3: {"name": "example", "balance": 300, "active": True, "description": None}
}

def test_health_status():
    response = client.get("/health")
    assert response.status_code == 200
    assert response.json() == {"status": True}

def test_account_creation():
    response = client.post("/accounts/4", json={"name": "new account name", "balance": 350000})
    assert response.status_code == 201
    assert response.json() == {
        "name": "new account name",
        "balance": 350000,
        "active": True,
        "description": None
    }

def test_existent_account_creation():
    response = client.post("/accounts/4", json={"name": "yet another account", "balance": 1})
    assert response.status_code == 409
    assert response.json() == {
        "detail": "Account exists"
    }

def test_account_removal():
    response = client.delete("/accounts/4")
    assert response.status_code == 200
    assert response.json() == True

def test_nonexistent_account_removal():
    response = client.delete("/accounts/4")
    assert response.status_code == 404
    assert response.json() == {"detail": "Account not found"}

def test_account_read():
    response = client.get("/accounts/1")
    assert response.status_code == 200
    assert response.json() == {
        "name": "joe",
        "balance": 500,
        "active": True,
        "description": None
    }

def test_nonexistent_account_read():
    response = client.get("/accounts/4")
    assert response.status_code == 404
    assert response.json() == {"detail": "Account not found"}


app/main.py
import asyncio
from typing import Dict, Optional, Union
from fastapi import FastAPI, HTTPException, Request
from pydantic import BaseModel

class Account(BaseModel):
    name: str
    description: Optional[str] = None
    balance: float
    active: bool = True

accounts = dict()

accounts = {
    1: {"name": "joe", "balance": 500, "active": True, "description": None},
    2: {"name": "smith", "balance": 400, "active": True, "description": None},
    3: {"name": "example", "balance": 300, "active": True, "description": None}
}

###################
# Metadata and Docs
###################

description = """
Features:

* Check health status of the application 🚀
* Create account
* Delete account
* Read account

"""

tags_metadata = [
    {
        "name": "accounts",
        "description": "Operations with accounts. Basic error handling has been implemented."
    },
    {
        "name": "health",
        "description": "Health checks that can be used to verify application status."
    }
]

app = FastAPI(title="Account Management",
    description=description,
    version="1",
    terms_of_service="https://www.example.org",
    contact={
        "name": "<Contact Name>",
        "url": "https://www.example.org",
        "email": "<Contact Email>",
    },
    license_info={
        "name": "<License Link>",
        "url": "https://www.example.org",
    },
    openapi_tags=tags_metadata
)

###########
# Functions
###########

async def get_account(account_id: int) -> Optional[Account]:
    if account_id in accounts:
        return accounts[account_id]
    else:
        return None

async def add_account(account_id: int, account: Account) -> Optional[Account]:
    if account_id in accounts:
        return None
    else:
        accounts[account_id] = account.dict()
        return accounts[account_id]

async def delete_account(account_id: int) -> Optional[bool]:
    if account_id in accounts:
        del accounts[account_id]
        return True
    else:
        return None

@app.get("/health", status_code=200, tags=["health"])
async def get_health(request: Request) -> Union[Optional[Dict], HTTPException]:
    return {"status": True}

@app.get("/accounts/{account_id}", status_code=200, tags=["accounts"])
async def read_account(account_id: int):
    res = await get_account(account_id)
    if res is None:
        raise HTTPException(status_code=404, detail="Account not found")
    else:
        return res

@app.post("/accounts/{account_id}", status_code=201, tags=["accounts"])
async def create_account(account_id: int, account: Account):
    res = await add_account(account_id, account)
    if res is None:
        raise HTTPException(status_code=409, detail="Account exists")
    else:
        return res

@app.delete("/accounts/{account_id}", status_code=200, tags=["accounts"])
async def remove_account(account_id: int):
    res = await delete_account(account_id)
    if res is None:
        raise HTTPException(status_code=404, detail="Account not found")
    else:
        return res
